### ISSUE TITLE
systemd-udevd should lock partitions after locking the parent devices

### ISSUE THREAD

**Issue by DemiMarie** [systemd/systemd#35244]
### Component

systemd-udevd

### Is your feature request related to a problem? Please describe

Tools like mkswap, mkfs, and fsck should not need to access the parent device, and ideally would not have access to them at all.  Furthermore, while most operations on partitions are cheap, some (such as encrypting a whole device) are expensive, so operations on different partitions should be able to happen in parallel.  Also, opening the parent device requires each tool to reimplement nontrivial code, whereas systemd-udevd can implement partition locking only once.

### Describe the solution you'd like

systemd-udevd should lock partitions, not just the parent devices.  There must be a defined order to avoid deadlocks.  I recommend locking the parent device first.

### Describe alternatives you've considered

Requiring each tool to find the parent device, which is slow and in some cases might require additional privileges.

**Comment by yuwata** [comment]
I do not think it is hard to check the parent device is locked or not, but we also provide `udevadm lock` command, so the commands like mkswap or so do not necessary to check by themselves if the whole block device is locked or not.

**Comment by poettering** [comment]
The problem is that various properties of partition block devices (i.e. ID_PART_ENTRY* and so on) are actually read from the main block device (because the are stored in the part table, not in the partitions), hence to get a reasonably atomic view of the device it only really makes sense to lock the whole thing for udev.

**Comment by DemiMarie** [comment]
Tools like mkfs, mkswap, and fsck don’t modify the partition table, so it doesn’t matter whether udev reads these properties before or after the tool runs.

**Comment by poettering** [comment]
Well, but those are not the only tools in the world. typically partitions are not considered just by their contents only but by their whole metadata which includes the GPT fields...

The problem with taking two locks is that we'd have to extend the locking protocol in an incompatible way, and it gets a lot more complex, because suddenly we have to dictate clearly the locking order, otherwise people will sloppily run into ABBA problems.

I don't think finding your containing block device is really that hard... 

An option might be to extend sd-device to add a helper that does this.

i.e. sd_device_lock() or so, which validates we are talking about a block device, then finds the right parent device, and then returns a lock fd.

**Comment by DemiMarie** [comment]
> Well, but those are not the only tools in the world. typically partitions are not considered just by their contents only but by their whole metadata which includes the GPT fields...

udev _does_ care about the partition metadata, but I suspect it is the exception here, not the norm.  mkfs, mkswap, and fsck definitely have no business caring about it.

> The problem with taking two locks is that we'd have to extend the locking protocol in an incompatible way, and it gets a lot more complex, because suddenly we have to dictate clearly the locking order, otherwise people will sloppily run into ABBA problems.

Which existing applications would this break?  Locking order is easy: lock the parent device first, then the partitions.

> I don't think finding your containing block device is really that hard...

1. It’s a lot harder than calling `flock()` on a file descriptor one already has open.
2. It serializes execution of what should be completely unrelated operations.  If I’m wiping an HDD partition I do _not_ want to block execution of operations on other partitions.
3. It requires the tool to care about something it really has no business caring about.  Some tools don’t even care that the file they are given is a block device, and none of them care whether they are given a partition or not.  They operate on a single file and that’s it.  Ideally `fsck` would not even have access to anything else!

It’s easy for udev to find the right block device because it already has all the infrastructure to do that, but it is much harder for other tools.  I wrote an implementation myself and it was not simple.
 
> An option might be to extend sd-device to add a helper that does this.
> 
> i.e. sd_device_lock() or so, which validates we are talking about a block device, then finds the right parent device, and then returns a lock fd.

Not every tool uses libsystemd.
